### DISCUSSION TITLE
Proposal: Searched switch statement

### DISCUSSION THREAD

**Discussion by wdnijdam** [dotnet/csharplang#8235] [Language Ideas]
# Searched switch statement

* [x] Proposed
* [ ] Prototype: Not Started
* [ ] Implementation: Not Started
* [ ] Specification: Not Started

## Summary
[summary]: #summary

Create a more convenient way to write a searched set of conditions resulting in a single result.

## Motivation
[motivation]: #motivation

The current alternatives of writing a searched switch are verbose and do not clearly indicate to the reader that a result is generated from a list of conditions and expressions

## Detailed design
[design]: #detailed-design

A search switch consists of a series conditions with expressions. The expression for first condition met should be returned. I should be clear for the reader that one answer is returned. The conditions are evaluated top to bottom. Conditions after the met condition are not executed.

Proposed syntax:

```
return switch
{
  when Cond1 => Expr1,
  when Cond2 => Expr2
  => Expr3
};
```

## Drawbacks
[drawbacks]: #drawbacks

Possible confusion with other switch expressions syntaxes. In this case the switch is executed without input object and may not be clear.

## Alternatives
[alternatives]: #alternatives

Use of current switch with constant input, having additional verbose:

```
return 0 switch
{
  _ when Cond1 => Expr1,
  _ when Cond2 => Expr2
  _ => Expr3
};
```

Use of listed if statements. It is not very clear for the reader that al branches contribute to the same result. It is only usable for method results bailing out with if statements. And it is fairly verbose on multiple lines

```
if (Cond1)
{
     return Expr1;
}
 
if (Cond2)
{
     return Expr2;
}
 
return Expr3;
```

Using ternary if statements in a cascade. This is hard to read and error prone.

```
return Cond1
     ? Expr1
     :  Cond2
             ? Expr2
             : Expr3;
```

or layout like (may be hard to read for longer conditions and expressions). It is unreadable when longer statements and more conditions are added.

```
return 
     Cond1 ? Expr1 :  
     Cond2 ? Expr2 : 
     Expr3;
```

Use of syntax like in TSQL:

```
return case
   when Cond1 then Expr1
   when Cond2 then Expr2
   else Expr3;
```

## Unresolved questions
[unresolved]: #unresolved-questions

**Comment by colejohnson66** [comment]
#8109

**Comment by 333fred** [comment]
As mentioned in our README, all new proposals need to start as discussions unless a member of the LDM directs you to do otherwise.
